File: 104_bk/21261212/cpp/pro05.cpp
/*
Please (1) explain the meaning of forward declaration (reduce dependency about header file) (2) Please rewrite the sample code with forward declaration.

//file C.h
#include "A.h"
#include "B.h"
class C 
{
    A* a;
    B b;
};

Ans:

gcc --version

gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

Under My Ubuntu 22.04.3 LTS VM, execute "gcc -o pro05 pro05.cpp" to test codes.

*/

//file C.h
class A;
class B;
class C 
{
    A* a;
    B b;
};

File: 104_bk/21261212/csharp/pro05.cs
/*
	Please consider the following code.
using System;
public class Singleton
{
   private static Singleton instance;
   private Singleton() {}
   public static Singleton Instance
   {
      get 
      {
         if (instance == null)
         {
            instance = new Singleton();
         }
         return instance;
      }
   }
}
	
(1) What is the problem with the code above. Please explain it briefly.
(2) Please rewrite it to solve the problem.

Ans:

*/